yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -222,7 +222,9 @@ private[spark] class ApplicationMaster(
 
         if (!unregistered) {
           // we only want to unregister if we don't want the RM to retry
-          if (finalStatus == FinalApplicationStatus.SUCCEEDED || isLastAttempt) {
+          if (finalStatus == FinalApplicationStatus.SUCCEEDED ||
+            exitCode == ApplicationMaster.EXIT_EARLY ||
+            exitCode == ApplicationMaster.EXIT_EXCEPTION_USER_CLASS || isLastAttempt) {
```

Comment:
@tgravescs 
What about when `SignalLogger.scala`
`log.error("RECEIVED SIGNAL " + signal.getNumber() + ": SIG" + signal.getName())`
when we kill the app using yarn kill we get this:
_ERROR ApplicationMaster: RECEIVED SIGNAL 15: SIGTERM_
can we use it to trigger cleanup?